### PR TITLE
Sanitize and bind "timeperiod" configuration queries

### PR DESCRIPTION
**Pull request by hyahiaoui-ext**
## Description

Sanitizing and binding timeperiod queries to avoid any sql injections attacks in the future + cleaning up legacy code.

**Fixes** # MON-13314

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [x] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [x] 21.04.x
- [x] 21.10.x
- [x] 22.04.x
- [x] 22.10.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 1]
```diff
@@ -475,3 +465,104 @@ function testTemplateLoop($value)
...
+ * Creates time period and returns id.
+ *
+ * @param array $params
+ * @return int
+ */
+function createTimePeriod(array $params): int
+{
+    global $pearDB;
+
+    $pearDB->query("INSERT INTO timeperiod VALUES (" . $params['values'] . ")");
```
Comment (jeremyjaouen) [inline www/include/configuration/configObject/timeperiod/DB-Func.php:509]: this query should also be sanitized

File: www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 2]
```diff
@@ -475,3 +465,104 @@ function testTemplateLoop($value)
...
+    return $pearDB->lastInsertId();
```
Comment (jeremyjaouen) [inline www/include/configuration/configObject/timeperiod/DB-Func.php:510]:
```suggestion
    return (int) $pearDB->lastInsertId();
```

If you cast here, you don't have to cast (int) $params['tp_id'] in your other following methods,
then your @return int in your phpdoc will be correct :), as lastInsertId() return a string

File: www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 3]
```diff
@@ -475,3 +465,104 @@ function testTemplateLoop($value)
...
+}
+
+/**
+ * Creates time periods exclude relations
+ *
+ * @param array $params
+ * @return void
```
Comment (jeremyjaouen) [inline www/include/configuration/configObject/timeperiod/DB-Func.php:517]:
```suggestion
```

File: www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 4]
```diff
@@ -475,3 +465,104 @@ function testTemplateLoop($value)
...
+    $statement->bindValue(':tp_id', (int) $params['tp_id'], \PDO::PARAM_INT);
+    $statement->bindValue(':timeperiod_id', (int) $params['timeperiod_id'], \PDO::PARAM_INT);
+    $statement->execute();
+}
+
+/**
+ * Creates time periods include relations
+ *
+ * @param array $params
+ * @return void
```
Comment (jeremyjaouen) [inline www/include/configuration/configObject/timeperiod/DB-Func.php:536]:
```suggestion
```

File: www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 5]
```diff
@@ -475,3 +465,104 @@ function testTemplateLoop($value)
...
+    $statement->bindValue(':tp_id', (int) $params['tp_id'], \PDO::PARAM_INT);
+    $statement->bindValue(':timeperiod_id', (int) $params['timeperiod_id'], \PDO::PARAM_INT);
+    $statement->execute();
+}
+
+/**
+ * Creates time periods exceptions
+ *
+ * @param array $params
+ * @return void
```
Comment (jeremyjaouen) [inline www/include/configuration/configObject/timeperiod/DB-Func.php:555]:
```suggestion
```

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

File: www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 6]
```diff
@@ -475,3 +465,126 @@ function testTemplateLoop($value)
...
+    $statement->execute();
+}
+
+/**
+ * Converts a string null value to a null.
+ *
+ * @param string $value
+ * @return string|null
+ */
+function notConvertNullToString(string $value)
```
Comment (kduret) [inline www/include/configuration/configObject/timeperiod/DB-Func.php:587]:
```suggestion
function convertNullStringToNull(string $value)
```

File: www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 7]
```diff
@@ -475,3 +465,126 @@ function testTemplateLoop($value)
...
+ * Creates time period and returns id.
+ *
+ * @param array $params
+ * @return int
+ */
+function createTimePeriod(array $params): int
+{
+    global $pearDB;
+
+    $valuesExploded = array_map('notConvertNullToString', explode(', ', $params['values']));
```
Comment (kduret) [inline www/include/configuration/configObject/timeperiod/DB-Func.php:509]:
```suggestion
    $valuesExploded = array_map('convertNullStringToNull', explode(', ', $params['values']));
```

**Review by kduret — COMMENTED** [review]
I think it is a better approach to replace `"NULL"` by `null` directly on variable assignment, it will avoid to create `convertNullStringToNull`

**Comment by hyahiaoui-ext** [comment]
> I think it is a better approach to replace `"NULL"` by `null` directly on variable assignment, it will avoid to create `convertNullStringToNull`

@kduret i know it's just here a bit complicated because it's mixed with strings and when you assign null directly to the string it's replaced by voids and then the query doesn't understand them ^^.

File: www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 8]
```diff
@@ -85,15 +85,20 @@ function testTPExistence($name = null)
...
+        htmlentities($centreon->checkIllegalChar($name), ENT_QUOTES, "UTF-8"),
+        \PDO::PARAM_STR
+    );
+    $statement->execute();
+    $tp = $statement->fetch(\PDO::FETCH_ASSOC);
     #Modif case
-    if ($dbResult->rowCount() >= 1 && $tp["tp_id"] == $id) {
+    if ($statement->rowCount() >= 1 && $tp["tp_id"] == $id) {
         return true;
     } #Duplicate entry
```
Comment (kduret) [inline www/include/configuration/configObject/timeperiod/DB-Func.php:100]:
```suggestion
```

File: www/include/configuration/configObject/timeperiod/DB-Func.php [hunk 9]
```diff
@@ -85,15 +85,20 @@ function testTPExistence($name = null)
...
-    elseif ($dbResult->rowCount() >= 1 && $tp["tp_id"] != $id) {
+    elseif ($statement->rowCount() >= 1 && $tp["tp_id"] != $id) {
```
Comment (kduret) [inline www/include/configuration/configObject/timeperiod/DB-Func.php:101]:
```suggestion
    } elseif ($statement->rowCount() >= 1 && $tp["tp_id"] != $id) { #Duplicate entry
```

**Review by jeremyjaouen — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**